jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java

```diff
@@ -873,7 +873,10 @@ public void onCompleted()
             LOG.debug("onCompleted for {} written={}", getRequest().getRequestURI(), getBytesWritten());
 
         if (_requestLog != null)
+        {
+            _request.onRequestLog();
```

Comment:
Could we not just move the call to `_request.onCompleted()` to before this line and then avoid creating another event and callback to the request?

Reply:
I believe `onCompleted()` is too late, by that point the request / response has been recycled already.

Reply:
We also don't want all of the unnecessary steps (like the restoring) to occur if there's no RequestLog declared / active.

Reply:
@joakime I believe that we can move the call to `onCompleted` to be just before the request log steps.  I think it is at least worth a try.   It makes sense to not log the request until it's handling is completed.

Reply:
Moving the fixing of Request to Request.onComplete() didn't work for the logback test suite.

Moving the fixing of the Request (contentParamsExtracted and status) to Request.onCompleted() breaks logging for session information (they get auth[nz] details out from sessions), and multipart information (file sizes), etc.

Currently, with the PR like it sits here, the RequestLog.log() occurs during HttpChannel.onComplete(), and before Channel idle timeout reset, and Request.onCompleted().
Turns out this location is the perfect place for RequestLog.log() to make sense.

We need another event before Request.onComplete() to fixate the Request/Response.
Or move the exiting behaviors in Request.onComplete() to a new event like Request.recycle() so we can use Request.onComplete() for RequestLog behaviors.
The two events with a RequestLog.log() in between is still needed.

Stated in a different way.

1. HttpChannel.onComplete()
2. Fixate Request/Response.
3. RequestLog.log().
4. Cleanup Request.

Reply:
I have a potential solution.
At least one that logback tests seem happy with.